sync/src/relayer/compact_block_process.rs

```diff
@@ -69,35 +72,50 @@ impl<'a, CS: ChainStore + 'static> CompactBlockProcess<'a, CS> {
                 );
                 return Ok(());
             }
-        } else {
-            debug!(target: "relay", "UnknownParent: {:#x}, send_getheaders_to_peer({})", block_hash, self.peer);
-            self.relayer.shared.send_getheaders_to_peer(
-                self.nc.as_ref(),
-                self.peer,
-                self.relayer.shared.lock_chain_state().tip_header(),
-            );
-            return Ok(());
+        }
+
+        if let Some(flight) = self
+            .relayer
+            .peers
+            .blocks_inflight
+            .read()
+            .inflight_state_by_block(&block_hash)
+        {
+            if flight.peers.contains(&self.peer) {
+                debug!(target: "relay", "discard already in-flight compact block {:x}", block_hash);
+                return Ok(());
+            }
         }
 
         // The new arrived has greater difficulty than local best known chain
         let mut missing_indexes: Vec<usize> = Vec::new();
         {
             // Verify compact block
             let mut pending_compact_blocks = self.relayer.state.pending_compact_blocks.lock();
-            if pending_compact_blocks.get(&block_hash).is_some()
+            if pending_compact_blocks
+                .get(&block_hash)
+                .map(|(_, peers_set)| peers_set.contains(&self.peer))
+                .unwrap_or(false)
                 || self.relayer.shared.get_block(&block_hash).is_some()
             {
-                debug!(target: "relay", "already processed compact block {:x}", block_hash);
+                debug!(target: "relay", "discard already pending compact block {:x}", block_hash);
                 return Ok(());
             } else {
+                let fn_get_pending_header = {
+                    |block_hash| {
+                        pending_compact_blocks
+                            .get(&block_hash)
+                            .map(|(compact_block, _)| compact_block.header.to_owned())
+                    }
+                };
                 let resolver = HeaderResolverWrapper::new(
                     &compact_block.header,
                     self.relayer.shared.shared().to_owned(),
                 );
                 let header_verifier = HeaderVerifier::new(
                     CompactBlockMedianTimeView {
                         anchor_hash: compact_block.header.hash(),
-                        pending_compact_blocks: &pending_compact_blocks,
+                        fn_get_pending_header: Box::new(fn_get_pending_header),
```

Comment:
I don't understand why change to use closure, could you explain the detail issue here?

Reply:
The structure of `pending_compact_blocks` has been changed, 
from 

```
Mutex<FnvHashMap<H256, CompactBlock>
```

to
 
```
Mutex<FnvHashMap<H256, (CompactBlock, FnvHashSet<PeerIndex>)>>
```

The `CompactBlockMedianTimeView` don't care the second element `FnvHashSet<PeerIndex>`. So I use closure to wrap it.

Another personal reason is, I think passing closure is more graceful, and I don't see any overhead here.

Reply:
`CompactBlockMedianTimeView`  is a private struct, I would like to keep it simple instead of reusable, how about using a type alias in this case? 

```
pub type PendingCompactBlocks = FnvHashMap<H256, (CompactBlock, FnvHashSet<PeerIndex>)>
```

```
struct RelayState {
    pending_compact_blocks: Mutex<PendingCompactBlocks>,
    ...
}

struct CompactBlockMedianTimeView {
    pending_compact_blocks: &'a PendingCompactBlocks,
    ...
}
```

Reply:
I do have already refactored it in other PR, but it has to wait for its prior PRs merged:
https://github.com/keroro520/ckb/pull/6
https://github.com/keroro520/ckb/pull/5
https://github.com/keroro520/ckb/pull/4

I want to leave it later.